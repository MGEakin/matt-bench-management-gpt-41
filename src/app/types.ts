import { z } from "zod";

export const EmployeeSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
  title: z.string(),
  studio: z.string(),
  practice: z.string(),
  region: z.string(),
  location: z.string(),
  skill_level: z.string(),
  current_assignment: z.string().optional(),
});

export type Employee = z.infer<typeof EmployeeSchema>;
