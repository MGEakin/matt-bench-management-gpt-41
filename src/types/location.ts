import { z } from "zod";

export const LocationSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Location = z.infer<typeof LocationSchema>;
