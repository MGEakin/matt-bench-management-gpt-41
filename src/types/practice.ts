import { z } from "zod";

export const PracticeSchema = z.object({
  id: z.number(),
  name: z.string(),
});

export type Practice = z.infer<typeof PracticeSchema>;
