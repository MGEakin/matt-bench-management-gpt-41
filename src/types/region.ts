import { z } from "zod";

export const RegionSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Region = z.infer<typeof RegionSchema>;
