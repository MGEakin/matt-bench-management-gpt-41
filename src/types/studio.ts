import { z } from "zod";

export const StudioSchema = z.object({
  id: z.number().optional(),
  name: z.string(),
});

export type Studio = z.infer<typeof StudioSchema>;
