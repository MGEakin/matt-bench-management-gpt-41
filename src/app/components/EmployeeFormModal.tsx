"use client";
import React, { useState, useEffect } from "react";
import { Employee } from "../../types/employee";
import { Practice } from "../../types/practice";
import { fetchPractices } from "../../lib/fetchPractices";

interface EmployeeFormModalProps {
  open: boolean;
  onClose: () => void;
  onSave: (employee: Omit<Employee, "id">, id?: number) => void;
  initialData?: Employee | null;
}

const defaultEmployee: Omit<Employee, "id"> = {
  name: "",
  title: "",
  studio: "",
  practice: "",
  region: "",
  location: "",
  skill_level: "",
  current_assignment: "",
};

const EmployeeFormModal: React.FC<EmployeeFormModalProps> = ({
  open,
  onClose,
  onSave,
  initialData,
}) => {
  const [form, setForm] = useState<Omit<Employee, "id">>(defaultEmployee);
  const [practices, setPractices] = useState<Practice[]>([]);
  const [loadingPractices, setLoadingPractices] = useState(true);
  const [errorPractices, setErrorPractices] = useState<string | null>(null);

  useEffect(() => {
    if (initialData) {
      const { id, ...rest } = initialData;
      setForm(rest);
    } else {
      setForm(defaultEmployee);
    }
  }, [initialData, open]);

  useEffect(() => {
    let mounted = true;
    setLoadingPractices(true);
    fetchPractices()
      .then((data) => {
        if (mounted) {
          setPractices(data);
          setLoadingPractices(false);
        }
      })
      .catch((err) => {
        if (mounted) {
          setErrorPractices(err.message);
          setLoadingPractices(false);
        }
      });
    return () => {
      mounted = false;
    };
  }, []);

  if (!open) return null;

  return (
    <div className="fixed inset-0 bg-black bg-opacity-40 flex items-center justify-center z-50">
      <div
        className="bg-white dark:bg-gray-900 p-6 rounded-lg shadow-lg w-full max-w-lg"
        role="dialog"
        aria-modal="true"
      >
        <h2 className="text-xl font-bold mb-4">
          {initialData ? "Edit Employee" : "Add Employee"}
        </h2>
        <form
          className="grid grid-cols-2 gap-4"
          onSubmit={(e) => {
            e.preventDefault();
            onSave(form, initialData?.id);
          }}
        >
          {/* Render all fields except practice as text inputs */}
          {Object.keys(defaultEmployee)
            .filter((key) => key !== "practice")
            .map((key) => (
              <div key={key} className="col-span-2 sm:col-span-1">
                <label className="block text-sm font-medium mb-1 capitalize">
                  {key.replace("_", " ")}
                </label>
                <input
                  className="w-full border rounded px-2 py-1 dark:bg-gray-800 dark:text-white"
                  type="text"
                  value={form[key as keyof typeof form] || ""}
                  onChange={(e) => setForm({ ...form, [key]: e.target.value })}
                  required={key !== "current_assignment"}
                  placeholder={key
                    .replace("_", " ")
                    .replace(/\b\w/g, (l) => l.toUpperCase())}
                />
              </div>
            ))}
          {/* Practice dropdown */}
          <div className="col-span-2 sm:col-span-1">
            <label className="block text-sm font-medium mb-1">Practice</label>
            {loadingPractices ? (
              <div className="text-gray-500">Loading practices...</div>
            ) : errorPractices ? (
              <div className="text-red-500">{errorPractices}</div>
            ) : (
              <select
                className="w-full border rounded px-2 py-1 dark:bg-gray-800 dark:text-white"
                value={form.practice}
                onChange={(e) => setForm({ ...form, practice: e.target.value })}
                required
              >
                <option value="" disabled>
                  Select a practice
                </option>
                {practices.map((practice) => (
                  <option key={practice.id} value={practice.name}>
                    {practice.name}
                  </option>
                ))}
              </select>
            )}
          </div>
          <div className="col-span-2 flex justify-end gap-2 mt-4">
            <button
              type="button"
              className="px-4 py-2 rounded bg-gray-300 dark:bg-gray-700"
              onClick={onClose}
            >
              Cancel
            </button>
            <button
              type="submit"
              className="px-4 py-2 rounded bg-blue-600 text-white"
            >
              {initialData ? "Update" : "Create"}
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default EmployeeFormModal;
