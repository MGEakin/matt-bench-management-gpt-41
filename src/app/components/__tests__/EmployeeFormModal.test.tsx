import React from "react";
import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import EmployeeFormModal from "../EmployeeFormModal";
import { Employee } from "../../types";

import "@testing-library/jest-dom";
import { create } from "react-test-renderer";

describe("EmployeeFormModal", () => {
  const onClose = vi.fn();
  const onSave = vi.fn();
  const initialData: Employee = {
    id: 1,
    name: "Alice Smith",
    title: "QA Lead",
    studio: "Studio A",
    practice: "Testing",
    region: "Americas",
    location: "New York",
    skill_level: "Senior",
    current_assignment: "Project X",
  };

  it("renders modal with initial data for edit", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={onClose}
        onSave={onSave}
        initialData={initialData}
      />,
    );
    expect(screen.getByDisplayValue("Alice Smith")).toBeInTheDocument();
    expect(screen.getByText("Edit Employee")).toBeInTheDocument();
  });

  it("calls onClose when Cancel is clicked", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={onClose}
        onSave={onSave}
        initialData={initialData}
      />,
    );
    fireEvent.click(screen.getByText("Cancel"));
    expect(onClose).toHaveBeenCalled();
  });

  it("calls onSave when form is submitted", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={onClose}
        onSave={onSave}
        initialData={initialData}
      />,
    );
    fireEvent.click(screen.getByText("Update"));
    expect(onSave).toHaveBeenCalled();
  });

  it("renders form fields", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={() => {}}
        onSave={() => {}}
        initialData={initialData}
      />,
    );
    expect(screen.getByDisplayValue("Alice Smith")).toBeInTheDocument();
    expect(screen.getByDisplayValue("QA Lead")).toBeInTheDocument();
  });

  it("renders empty form for new employee", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={() => {}}
        onSave={() => {}}
        initialData={undefined}
      />,
    );
    expect(screen.getByText("Add Employee")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Name")).toHaveValue("");
  });

  it("matches snapshot", () => {
    const tree = create(
      <EmployeeFormModal
        open={true}
        onClose={() => {}}
        onSave={() => {}}
        initialData={initialData}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has accessible dialog", () => {
    render(
      <EmployeeFormModal
        open={true}
        onClose={() => {}}
        onSave={() => {}}
        initialData={initialData}
      />,
    );
    expect(screen.getByRole("dialog")).toBeInTheDocument();
  });
});
