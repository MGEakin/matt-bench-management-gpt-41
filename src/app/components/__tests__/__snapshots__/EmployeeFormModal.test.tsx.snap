// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmployeeFormModal > matches snapshot 1`] = `null`;
